calyx-opt/src/passes/well_formed.rs

```diff
@@ -191,16 +191,24 @@ where
     Ok(())
 }
 
-fn same_type(proto_out: &CellType, proto_in: &CellType) -> CalyxResult<()> {
-    if proto_out != proto_in {
-        Err(Error::malformed_control(format!(
-            "Unexpected type for ref cell. Expected `{}`, received `{}`",
-            proto_out.surface_name().unwrap(),
-            proto_in.surface_name().unwrap(),
-        )))
-    } else {
-        Ok(())
+/// Returns true if cell_in is a subtype of the output cell.
+// XXX(nathanielnrn): I think this is incomplete, technically we should also
+// be iterating over ref cells of each cell and that they are contravariant?
+// However cells don't have this information so we'd need to extract it from `comps` I guess?
```

Comment:
I think this is a good point. I suppose my instinct would be to move ahead anyway and create an issue w/r/t tightening this up? It seems to require some representation changes, as you mention, so it is worth some thought about how to make this convenient.

Reply:
#2123 discusses this.